File: cpp_version/binary_trees/how_high/how_high.cpp
#include <algorithm>
#include <cmath>
#include <iostream>
#include <limits.h>
#include <math.h>
#include <queue>
#include <stack>
#include <string>
#include <tuple>
#include <unordered_map>
#include <unordered_set>
#include <vector>

// how high
// Write a function, howHigh, that takes in the root of a binary tree.
// The function should return a number representing the height of the tree.

// The height of a binary tree is defined as the maximal number of edges from the root node to any leaf node.

// If the tree is empty, return -1.

class Node {
public:
    std::string val;
    Node *left;
    Node *right;

    Node(std::string initialVal) {
        val = initialVal;
        left = nullptr;
        right = nullptr;
    }
};

int howHigh(Node *root) {
    // todo


    return 1 
}

int main() {

    Node a("a");
    Node b("b");
    Node c("c");
    Node d("d");
    Node e("e");
    Node f("f");

    a.left = &b;
    a.right = &c;
    b.left = &d;
    b.right = &e;
    c.right = &f;

    //      a
    //    /   \
//   b     c
    //  / \     \
// d   e     f

    std::cout << howHigh(&a) << std::endl; // -> 2
    // Node a("a");
    // Node b("b");
    // Node c("c");
    // Node d("d");
    // Node e("e");
    // Node f("f");
    Node g("g");

    a.left = &b;
    a.right = &c;
    b.left = &d;
    b.right = &e;
    c.right = &f;
    e.left = &g;

    //      a
    //    /   \
//   b     c
    //  / \     \
// d   e     f
    //    /
    //   g

    std::cout << howHigh(&a) << std::endl; // -> 3
    // Node a("a");
    // Node c("c");

    a.right = &c;
    a.left = nullptr;
    c.left = nullptr;
    c.right = nullptr;

    //      a
    //       \
//        c

    std::cout << howHigh(&a) << std::endl; // -> 1
    // Node a("a");
    a.left = nullptr;
    a.right = nullptr;

    //      a

    std::cout << howHigh(&a) << std::endl;      // -> 0
    std::cout << howHigh(nullptr) << std::endl; // -> -1
    return 0;
}

File: javascript_version/binary_trees/how_high/how_high.js
// how high
// Write a function, howHigh, that takes in the root of a binary tree. 
// The function should return a number representing the height of the tree.

// The height of a binary tree is defined as the maximal number of edges from the root node to any leaf node.

// If the tree is empty, return -1.

class Node {
    constructor (val) {
        this.val = val;
        this.left = null;
        this.right = null;
    }
}

const howHigh = (root) => {
    if (root === null) { return -1; }
    return 1 + Math.max(howHigh(root.left) , howHigh(root.right));
};

const a = new Node('a');
const b = new Node('b');
const c = new Node('c');
const d = new Node('d');
const e = new Node('e');
const f = new Node('f');

a.left = b;
a.right = c;
b.left = d;
b.right = e;
c.right = f;

//      a
//    /   \
//   b     c
//  / \     \
// d   e     f

console.log(howHigh(a)); // -> 2

const g = new Node('g');

a.left = b;
a.right = c;
b.left = d;
b.right = e;
c.right = f;
e.left = g

//      a
//    /   \
//   b     c
//  / \     \
// d   e     f
//    /
//   g

console.log(howHigh(a)); // -> 3


a.right = c;
a.left = null;
c.right = null;
c.left = null;
//      a
//       \
//        c

console.log(howHigh(a)); // -> 1
// const a = new Node('a');
a.right = null;
a.left = null;
//      a

console.log(howHigh(a)); // -> 0
console.log(howHigh(null)); // -> -1